### PR TITLE
util/filesort: Make sorting async

### PR DESCRIPTION
**Pull request by yanzhe-chen**
Optimize the input phase in filesort.

Use auxiliary workers to make sorting in input phase async.

### Environment Setup

* One machine with two 10-core Intel Xeon E5-2650 v3 processors, 64GB of DRAM.
* Each core has a private 32KB L1 cache and a private 256KB L2 cache, and all 10 cores on a single processor share a 24MB L3 cache.
* Hyper-threading disabled.

### Workload Setup

* 10 million rows, the key size and value size are both 8 bytes
* Buffer size is set to 500K
* Adjust the number of workers from 1 to 16

### Evaluation results

image

Compared to a single worker, using 16 workers can gain up to **8.14X** speedup.

### CODE REVIEW

File: util/filesort/filesort.go [hunk 1]
```diff
@@ -335,66 +402,73 @@ func (fs *FileSorter) fetchNextRow(index int) (*comparableRow, error) {
...
+				break
+			}
+		}
+		if assigned {
+			fs.tCounter = 0
+			break
+		} else {
+			// if all workers are busy, block 100ms and try again
+			time.Sleep(100 * time.Millisecond)
+			fs.tCounter++
```
Comment (coocood) [inline util/filesort/filesort.go]: Use `time.Duration` for total wait time would be more clear.

**Comment by coocood** [comment]
@yanzhe-chen 
Use a buffio reader and writer can improve file read and write performance.

File: util/filesort/filesort.go [hunk 2]
```diff
@@ -104,31 +106,53 @@ func (rh *rowHeap) Pop() interface{} {
...
+	maxRowSize int
 }
 
-// Builder builds a new FileSorter.
-type Builder struct {
-	sc      *variable.StatementContext
+// Worker sorts file asynchronously.
+type Worker struct {
+	ctx     *FileSorter
+	busy    bool
```
Comment (coocood) [inline util/filesort/filesort.go]: This variable is accessed by multiple goroutines, should use atomic method.

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@shenli @zimulala PTAL

**Comment by yanzhe-chen** [comment]
I have reenabled the internal sort when possible.

The condition which internal sort is possible is slightly different:

`# of input rows <= bufSize / nWorkers`

(Originally it is: `# of input rows <= bufSize`)

@coocood @shenli @zimulala PTAL

File: util/filesort/filesort.go [hunk 3]
```diff
@@ -425,90 +518,113 @@ func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
...
 
-	if fs.rowHeap.Len() > 0 {
-		im := heap.Pop(fs.rowHeap).(*item)
-		if fs.rowHeap.err != nil {
-			return nil, errors.Trace(fs.rowHeap.err)
-		}
+	fileName := w.ctx.getUniqueFileName()
 
-		row, err := fs.fetchNextRow(im.index)
+	outputFile, err = os.OpenFile(fileName, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
```
Comment (zimulala) [inline util/filesort/filesort.go]: Use `:=` here, remove line 572 and line 573.

File: util/filesort/filesort.go [hunk 4]
```diff
@@ -188,144 +221,198 @@ func (b *Builder) Build() (*FileSorter, error) {
...
 
+	ws := make([]*Worker, b.nWorkers)
+	for i := range ws {
+		ws[i] = &Worker{
+			keySize: b.keySize,
+			valSize: b.valSize,
+			rowSize: b.keySize + b.valSize + 1,
+			bufSize: b.bufSize / b.nWorkers,
+			buf:     make([]*comparableRow, 0, b.bufSize/b.nWorkers),
+			head:    make([]byte, 8),
```
Comment (zimulala) [inline util/filesort/filesort.go]: Please define a constant as the head length. We used it in many places.

File: util/filesort/filesort.go [hunk 5]
```diff
@@ -188,144 +221,198 @@ func (b *Builder) Build() (*FileSorter, error) {
...
+	fs.files = append(fs.files, fn)
+}
 
-	sort.Sort(fs)
-	if fs.err != nil {
-		return errors.Trace(fs.err)
+func (fs *FileSorter) closeAllFiles() error {
+	for _, fd := range fs.fds {
+		err := fd.Close()
+		if err != nil {
```
Comment (zimulala) [inline util/filesort/filesort.go]: If one fd close failed,  we need to close other fds.

File: util/filesort/filesort.go [hunk 6]
```diff
@@ -425,90 +518,113 @@ func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
...
+		binary.BigEndian.PutUint64(w.head, uint64(len(outputByte)-prevLen-8))
+		for i := 0; i < 8; i++ {
+			outputByte[prevLen+i] = w.head[i]
+		}
+	}
 
-		return im.value, nil
+	_, err = outputFile.Write(outputByte)
+	if err != nil {
+		w.err = err
```
Comment (zimulala) [inline util/filesort/filesort.go]: Add trace.

File: util/filesort/filesort.go [hunk 7]
```diff
@@ -425,90 +518,113 @@ func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
...
-		im := heap.Pop(fs.rowHeap).(*item)
-		if fs.rowHeap.err != nil {
-			return nil, errors.Trace(fs.rowHeap.err)
-		}
+	fileName := w.ctx.getUniqueFileName()
 
-		row, err := fs.fetchNextRow(im.index)
+	outputFile, err = os.OpenFile(fileName, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		w.err = err
```
Comment (zimulala) [inline util/filesort/filesort.go]: Add trace.

File: util/filesort/filesort.go [hunk 8]
```diff
@@ -425,90 +518,113 @@ func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
...
+	}
+	defer outputFile.Close()
+
+	for _, row := range w.buf {
+		prevLen = len(outputByte)
+		outputByte = append(outputByte, w.head...)
+		outputByte, err = codec.EncodeKey(outputByte, row.key...)
 		if err != nil {
-			return nil, errors.Trace(err)
+			w.err = err
```
Comment (zimulala) [inline util/filesort/filesort.go]: Add trace.

File: util/filesort/filesort.go [hunk 9]
```diff
@@ -425,90 +518,113 @@ func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
...
+			return
+		}
+		outputByte, err = codec.EncodeKey(outputByte, row.val...)
+		if err != nil {
+			w.err = err
```
Comment (zimulala) [inline util/filesort/filesort.go]: Add trace.

File: util/filesort/filesort.go [hunk 10]
```diff
@@ -425,90 +518,113 @@ func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
...
+			return
+		}
+		outputByte, err = codec.EncodeKey(outputByte, types.NewIntDatum(row.handle))
+		if err != nil {
+			w.err = err
```
Comment (zimulala) [inline util/filesort/filesort.go]: Add trace.

File: util/filesort/filesort.go [hunk 11]
```diff
@@ -425,90 +518,113 @@ func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
...
-	if !fs.fetched {
-		sort.Sort(fs)
-		if fs.err != nil {
-			return nil, errors.Trace(fs.err)
-		}
-		fs.fetched = true
+// Close terminates the input or output process and discards all remaining data.
+func (fs *FileSorter) Close() error {
+	if fs.closed {
+		return errors.New("FileSorter has been closed")
```
Comment (zimulala) [inline util/filesort/filesort.go]: It's better to return nil.

File: util/filesort/filesort_test.go [hunk 12]
```diff
@@ -180,7 +180,56 @@ func (s *testFileSortSuite) TestMultipleFiles(c *C) {
 	c.Assert(err, IsNil)
 
 	fsBuilder := new(Builder)
-	fs, err = fsBuilder.SetSC(sc).SetSchema(keySize, valSize).SetBuf(bufSize).SetDesc(byDesc).SetDir(tmpDir).Build()
+	fs, err = fsBuilder.SetSC(sc).SetSchema(keySize, valSize).SetBuf(bufSize).SetWorkers(1).SetDesc(byDesc).SetDir(tmpDir).Build()
+	c.Assert(err, IsNil)
```
Comment (zimulala) [inline util/filesort/filesort_test.go]: Add `defer fs.Close()`.

File: util/filesort/filesort.go [hunk 13]
```diff
@@ -425,90 +518,113 @@ func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
...
 	}
-	if fs.cursor < len(fs.buf) {
-		r := fs.buf[fs.cursor]
-		fs.cursor++
-		return r, nil
+	fs.wg.Wait()
+	err := fs.closeAllFiles()
+	if err != nil {
```
Comment (zimulala) [inline util/filesort/filesort.go]: If close all files failed, we need to clean up other information.

File: util/filesort/filesort.go (hunk 4 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed

File: util/filesort/filesort.go (hunk 5 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

File: util/filesort/filesort.go (hunk 13 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

File: util/filesort/filesort.go [hunk 14]
```diff
@@ -188,144 +221,198 @@ func (b *Builder) Build() (*FileSorter, error) {
...
 	var (
-		err  error
-		n    int
-		head = make([]byte, 8)
-		dcod = make([]types.Datum, 0, fs.keySize+fs.valSize+1)
+		err error
+		n   int
 	)
-	n, err = fs.fds[index].Read(head)
+	n, err = fs.fds[index].Read(fs.head)
```
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed

File: util/filesort/filesort.go (hunk 11 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

File: util/filesort/filesort.go (hunk 3 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

File: util/filesort/filesort_test.go (hunk 12 above)
Comment (yanzhe-chen) [inline util/filesort/filesort_test.go]: Fixed.

File: util/filesort/filesort.go (hunk 7 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

File: util/filesort/filesort.go (hunk 8 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

File: util/filesort/filesort.go (hunk 9 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

File: util/filesort/filesort.go (hunk 10 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

File: util/filesort/filesort.go (hunk 6 above)
Comment (yanzhe-chen) [inline util/filesort/filesort.go]: Fixed.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**